### ISSUE TITLE
Localhost identity server 4 setup - possible bug

### ISSUE THREAD

**Issue by PeterKottas** [FormidableLabs/react-native-app-auth#44]
Hello all. First I'd like to thank you for putting together this lib. It seems to be fairly fresh (91 stars at the time of writing) but it really feels like it will pick up. So far it's by far the best way to authenticate in react-native I found.

It's quite difficult for me to identify with 100% accuracy if this is a bug so I'll try to explain what I am doing.
I started off the example (copied over the app code) which works fine. I wanted to swap the demo server with my own running on my machine. So I swapped the issuer for my http://localhost:53721 and opened the reverse proxy via `adb reverse tcp:53721 tcp:53721` so that I can access it on my emulator (API 23 Android 6.0 (Google APIs)). I left the redirectUrl what it was before (redirectUrl: 'io.identityserver.demo:/oauthredirect') partially cause I am not exactly sure what to put there in case of redirect. 

What happens when I run the app and click Authorize, it stops for a second and then "Unfortunately, XXX has stopped. OK" Alert message. 

While I understand that I might need to fix the redirectUrl both in app and in build.gradle, it still shouldn't fail this critically IMHO. Especially when it's surrounded with try catch block. 

Am I doing something wrong or is this a bug? Any input much appreciated.

**Comment by kadikraman** [comment]
Hi @PeterKottas - are you running Identity Server 4 locally? The issuer should be the root url for Identity Server. You can verify this by opening http://localhost:53721/.well-known/openid-configuration in your browser - you should see a JSON response with a bunch of urls. This should also work in the browser of your emulator.

As for the `redirectUrl`, it is basically a deep link back into the application with the auth code following a successful authentication.

If you're setting up IdentityServer, you would define the redirect urls, e.g.

```
RedirectUris = new List<string>
{
   "com.myapp:/oauthredirect",
},
```

Where `com.myapp` is the the same configuration as set in your `build.gradle`. It is a convention for mobile apps to have the app name be a reverse of their url, so you'd usually replace `myapp` with whatever your app is called.

**Comment by brockallen** [comment]
> What happens when I run the app and click Authorize, it stops for a second and then "Unfortunately, XXX has stopped. OK" Alert message.

Is this from our android sample? i get this from time to time when I move the code around or make other changes. My sense is that it's from the fragility of xamarin and VS and how those all hang together. In short, I do usually need to delete all bin/obj, do a clean, and/or then rebuild from scratch.

**Comment by PeterKottas** [comment]
Thanks for quick rections guys!
@kadikraman 
1. Yep I can see the server and the automatic discovery doc on http://localhost:53721/.well-known/openid-configuration on my emulator so that works fine.
2. I now managed to fix the redirect scheme as well. Totally forgot about the reverse dns rule but that's now done.
@brockallen You are totally right. I was afraid it's just me but damn this thing is fragile. I started from create-react-native-app ejected right away, added react-native-web, configured some stuff, then added things from your example. So it's not exact fork but all major things appear to be in order.

Guys even with these setup, I still get the Unfortunately ... when I click authorize button. It's very strange. I am copying over config for my client in identity server just in case. Once again, it might be miss-configured, I admit to that but it still shouldn't fall down this drastically imho.
```
new Client
{
    ClientId = "guest",
    ClientName = "Guest Client",
    AllowedGrantTypes = GrantTypes.Implicit,
    AllowAccessTokensViaBrowser = true,
    RequireConsent = false,

    /*ClientSecrets =
    {
        new Secret("secret".Sha256())
    },*/

    RedirectUris = Configuration["Guest:Url"].Split(',').Select(item=>item + "oauthredirect").ToList(),
    PostLogoutRedirectUris = Configuration["Guest:Url"].Split(','),

    AllowedScopes =
    {
        IdentityServerConstants.StandardScopes.OpenId,
        IdentityServerConstants.StandardScopes.Profile,
        "GuestBell.Host.Backend"
    },
    AllowOfflineAccess = true,
    AccessTokenLifetime = 7200,
}
```

There's one thing left that I will have to check although it is almost definitely no the cause of this issue. Apparently identity server have some filtering setup that only allows web uri to be setup for redirect url. That means only http/https are supported while other schemes throw an exception and show error. The problem is app shuts down even before the popup appears so that shouldn't be a factor.

**Comment by kadikraman** [comment]
@PeterKottas here's the [IdentityServer4 config example](https://github.com/openid/AppAuth-iOS/blob/master/Examples/README-IdentityServer.md) form AppAuth-iOS (I know you're testing on Android, but the same configuration should work). Looks like there's a few things that are different from yours, notably `RequireClientSecret` and `AllowedGrantTypes`.

```
var client = new Client
{
    ClientId = "native.code",
    ClientName = "Native Client (Code with PKCE)",
    RequireClientSecret = false,
    
    RedirectUris = { "io.identityserver.demo:/oauthredirect" },

    AllowedGrantTypes = GrantTypes.Code,
    RequirePkce = true,
    AllowedScopes = { "openid", "profile" },
    AllowOfflineAccess = true
};
```
I agree that we should avoid app crashes if possible and provide a helpful error message instead. Let's figure out why it's crashing in the first place and then determine whether and how to protect against it.

Could you try modifying your client configuration to see if that clears the issue. It is possible that IdentityServer is returning a response that we are not handling properly.

**Comment by PeterKottas** [comment]
Silly me! I actually find this lib because I didn't want to use implicit as it's said not to be the right choice for mobile apps. Yet I failed to change it in config. Facepalm ... long week I guess. Anyways, it's now changed, I've enabled remote debugging for native app, set it to pause on caught exceptions. Yet the situation remains the same. 

I cannot publish the auth server but I would be able to publish the repo for mobile app, would that help?

**Comment by PeterKottas** [comment]
Some screenshots from android studio about sdk and build tools I've installed. 
![androidsdks](https://user-images.githubusercontent.com/10601911/36354378-baf01630-14d3-11e8-9c36-d7f9a5202837.jpg)
![sdktools](https://user-images.githubusercontent.com/10601911/36354379-bb0a0f68-14d3-11e8-978e-adad4adff3ef.jpg)

**Comment by PeterKottas** [comment]
My build.gradle in android/app
```
android {
    compileSdkVersion 25
    buildToolsVersion "26.0.2"
    defaultConfig {
        // More stuff
        manifestPlaceholders = [
            appAuthRedirectScheme: 'com.guestbell.guest'
        ]
        // More stuff
    }
    // More stuff
}
```
and the on in root of android
```

buildscript {
    repositories {
        jcenter()
        google()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.0.1'

        // NOTE: Do not place your application dependencies here; they belong
        // in the individual module build.gradle files
    }
}

allprojects {
    repositories {
        mavenLocal()
        jcenter()
        maven {
            // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
            url "$rootDir/../node_modules/react-native/android"
        }
    }
}
```

**Comment by PeterKottas** [comment]
I also managed to clone your example repo and run my auth server with the same results. So at least it seems I didn't do anything silly on the client side. I am wondering ... I think I've read somewhere there's a problem with app-auth-android without ssl. Can this be a factor?

**Comment by PeterKottas** [comment]
Ok I made solid progress! It seems the problem indeed is SSL is required for IDP provider. I managed to connect to my dev server which is setup with SSL cert (self signed so remember to upload to trusted certs otherwise there's hell to expect with cryptic error messages). All plays along nicely now! Hopefully this thread will be useful to some folks, I'll also open separate issues for things I found along the way. Thanks for your help guys!
